### ISSUE TITLE
dual functional-based maxErr/exceedance calc

### ISSUE THREAD

**Issue by benthestatistician** [markmfredrickson/optmatch#164]
Solving the MDF with a discretized distance leads in general to not-quite-optimal node prices $p'$, with a dual functional value $q(p')$ somewhat below $\sup_p q(p)$. The corresponding sum of matched distances, $d(x')$ say, is probably also somewhat above its minimum value: also $\sup_p q(p)$, by basic duality result.   Thus $d(x') - q(p')$ is an upper bound on the sub-optimality of $d(x')$. 
(For def. of dual functional $q()$, see Bertsekas '98, displays 4.2--4.4, pp.156-7).

This relates closely to evaluating complementary slackness (CS) in the problem w/ non-discretized distances (#160).  CS obtains iff $L(x',p') = min_x L(x, p') = q(p')$, where $L(,)$ is the Lagrangian of the non-discretized problem; see Bertsekas '98 ch.4. So it's natural to calculate this $q(p')$ as part of evaluating complementary slackness. 

(In "problem w/ non-discretized distances", we should using the double-precision distance that kicked off the exercise plus back-transformed stability increments.)

**Comment by benthestatistician** [comment]
Implemented for a single subproblem in 0bdc6f0.  Still to do: 

- [ ] Make `evaluate_*` functions work just as well with a (distance, MCFSolutions) pair combining multiple subproblems. 
- [ ] Review/consider tests to add.  (Mark's test from d6b6c3c is v. helpful, but I'd be surprised if that alone covers it.)
- [ ] ¿Align levels of the factors we join on, rather than relying on dplyr::inner_join() to convert to character?  (Some proposals for this in comments to #161.)

**Comment by benthestatistician** [comment]
When this is ready it can go into the summary.optmatch routine (#173).

**Comment by benthestatistician** [comment]
As noted in [this comment](https://github.com/markmfredrickson/optmatch/issues/166#issuecomment-751888537) to #166, for performance reasons we should also 

- [ ] consider whether there'd be performance savings to combining `evaluate_lagrangian()` and `evaluate_dual()` into a single routine. (Alternatively, `evaluate_primal()` and `evaluate_dual()`.)
- [ ] implement if/as appropriate. 

As discussed in the same note, it's presently unclear whether these functions should ultimately be invoked from within `solve_reg_fm_prob()`, as they are in the proj1-nodeprices branch at present (02e9a5a), or from `summary.optmatch()` (cf #173).

**Comment by benthestatistician** [comment]
Looking back on this work, I can't think of anything we actually need the Lagrangian value for.  Presently in the master branch it's being calculated and stored, as `lagrangian_value` in the subproblems table, but not used anywhere else.  I think we'd be better served to:
1.  Redefine the SubProblemInfo class to require a `primal_value` column rather than a `lagrangian_value` column;
2. In `solve_reg_fm_prob()`, invoke `evaluate_primal()` not `evaluate_lagrangian()`. 

Then we'd be able to get the regret immeidately.

I could well be forgetting something; speak up if you suspect that I am or may be.  @markmfredrickson @adamrauh

**Comment by adamrauh** [comment]
The commit over in my forked master branch [here](https://github.com/adamrauh/optmatch/commit/056ea1fa8d01479b7e6a9da5d279a9194b53a6e8) does a simple substitution of the primal for the Lagrangian. I can issue a pull request with this change + fixing the feasibility indicator [issue](https://github.com/markmfredrickson/optmatch/issues/178) if there are no objections @benthestatistician @markmfredrickson

**Comment by benthestatistician** [comment]
NB: redefining the SubProblemInfo class calls for updates in `setClass("SubProbInfo", <...>)` and `setValidity("SubProbInfo", <...>)`, over in MCFSolutions.R.  (And for running devtools::document() after that.)

**Comment by adamrauh** [comment]
To a similar end, I noticed that there are some tests that explicitly set `lagrangian_value` over in `test.MCFSolutions.R`. For example, this one here: https://github.com/markmfredrickson/optmatch/blob/4d9e3423adc095e591f3d5824d35c0e2c676d6e1/tests/testthat/test.MCFSolutions.R#L6

It seems like these are just kicking the tires on the constructor and related methods -- swapping in `primal_value` after making the appropriate changes elsewhere didn't break any tests. But, I'm not sure if there was special meaning to the values chosen or their association with the Lagrangian specifically. (My sense is no.)

(Sorry for the deleted comment. I realized there were some loose ends missed right after posting)

**Comment by benthestatistician** [comment]
You're right, it's just tire kicking.  No special meaning to the values chosen here. 

In replacing `lagrangian_value` with `primal_value`, I'd suggest ensuring that each subproblem's `primal_value` is no smaller than its `dual_value`. As it doesn't make sense to have a smaller primal than dual value, at least not for feasible problems with feasible solutions.
